src/components/implementation/no_interface/cbboot/quarantine.c

```diff
@@ -302,41 +320,57 @@ quarantine_fork(spdid_t spdid, spdid_t source)
 	/* mman: have to do this now so the memory maps are available. */
 	if (tot > SERVICE_SIZE) tot = SERVICE_SIZE + 3 * round_up_to_pgd_page(1) - tot;
 	else tot = SERVICE_SIZE - tot;
-	printd("Telling mman to fork(%d, %d, %d, %x, %d)\n", cos_spd_id(), source, d_spd, prev_map + PAGE_SIZE, tot);
 	
-	r = mman_fork_spd(cos_spd_id(), source, d_spd, prev_map + PAGE_SIZE, tot);
-	if (r) printc("Error (%d) in mman_fork_spd\n", r);
-
-#if 0
-	/* cbuf */
 	printd("Telling cbuf to fork(%d, %d, %d)\n", cos_spd_id(), source, d_spd);
-	r = cbuf_fork_spd(cos_spd_id(), source, d_spd);
-	if (r) printc("Error (%d) in cbuf_fork_spd\n", r);
-
-	/* TODO: valloc */
-#endif
+	cinfo_addr = cbuf_fork_spd(cos_spd_id(), source, d_spd, cinfo_cbid);
+	
+	/* fixup cinfo page */
+	struct cos_component_information *ci = (struct cos_component_information*) cinfo_addr;
+	ci->cos_this_spd_id = d_spd;
+	__boot_spd_set_symbs(src_hdr, d_spd, ci);
+	__boot_deps_save_hp(d_spd, ci->cos_heap_ptr);
+	
+	/* 
+	 * FIXME: set fault handlers and re-write caps 
+	 * Also a note that activate should come after cbuf_fork, since it expects memory to be copied over.
+	 */
+	if (cos_spd_cntl(COS_SPD_ACTIVATE, d_spd, src_hdr->ncap, 0)) BUG();
+	printc("Setting capabilities for %d\n", d_spd);
+	if (__boot_spd_caps(src_hdr, d_spd)) BUG();
+	printc("Setting capabilities done\n", d_spd);
+	
+	/* Fix send-side fork counters */
+	if (send_side_counters(source, d_spd, src_hdr)) BUG();
+	if (receive_side_counters(source, d_spd)) BUG();
 
 	quarantine_add_to_spd_map(source, d_spd);
 
-#if QUARANTINE_MIGRATE_THREAD
+#ifdef QUARANTINE_MIGRATE_THREAD
 	quarantine_migrate(cos_spd_id(), source, d_spd, d_thd);
-	printd("Waking up thread %d\n", d_thd);
 	if (d_thd) {
 		sched_quarantine_wakeup(cos_spd_id(), d_thd);
 	}
 #endif
 
 	/* TODO: should creation of boot threads be controlled by policy? */
-	printd("Creating boot threads in fork: %d\n", d_spd);
-	__boot_spd_thd(d_spd);
-
-done:
-	assert(0 == sched_curr_set_priority(p));
+	//printd("Creating boot threads in fork: %d\n", d_spd);
+	//if (__boot_spd_thd(d_spd)) BUG();
+	
+	int reset;
+	reset = sched_curr_set_priority(p);
+	assert(0 == reset);
 	printd("Forked %d -> %d\n", source, d_spd);
 	return d_spd;
-}
 
+error:
+	do { printc("Found an error\n"); } while (0);
```

Comment:
As I recall, gcc just would NOT compile this code saying that a printc could not come directly after a label. Most likely that's not true and I will just fix it.